### PR TITLE
Next version of the containers library

### PR DESCRIPTION
**Pull request by Hackerpilot**
Goals:
* More consistency in method names and argument order
* More opportunities to use the containers without accidentally using the GC
* Slightly better performance
* Ability to use more containers as output ranges

### CODE REVIEW

**Comment by Hackerpilot** [comment]
* Fixed a few previously-undiscovered bugs.
* Added documentation generation script.
* Generated documentation will be served out of the docs/ folder in the master branch once this is merged.
* Lots of documentation improvements and fixes.
* Added a better string hashing function.
* Improved space efficiency of the hash map.

**Comment by skl131313** [comment]
Is it a good idea to add the docs to the git repo? It looks like it is adding a ton of bloat, every function has its' own html file.

**Comment by wilzbach** [comment]
Yes, especially if we already use Travis to auto-generate the documentation and push it to the gh-pages branch: https://dlang-community.github.io/containers/index.html

**Comment by Hackerpilot** [comment]
I either didn't know or forgot that there was a CI job set up to keep the gh-pages branch up-to-date.

**Comment by Hackerpilot** [comment]
Fixes #63.

**Comment by Hackerpilot** [comment]
A version of the library before these changes is tagged as https://github.com/dlang-community/containers/releases/tag/v0.7.0.

File: src/containers/simdset.d [hunk 1]
```diff
@@ -110,6 +124,18 @@ version (D_InlineAsm_X86_64) struct SimdSet(T, Allocator = Mallocator)
 		return true;
 	}
 
+    /// ditto
+    bool opOpAssign(string op)(T item) if (op == "~")
```
Comment (Hackerpilot) [inline src/containers/simdset.d]: Indentation.

File: src/containers/slist.d [hunk 2]
```diff
@@ -64,28 +64,45 @@ struct SList(T, Allocator = Mallocator, bool supportGC = shouldAddGCRange!T)
...
 		return cast(ET) _front.value;
 	}
 
 	/**
-	 * Removes and returns the first item in the list.
+	 * Complexity: O(length)
+	 * Returns: the least recently inserted item.
+	 */
+	auto back(this This)() inout @property
+	{
```
Comment (Hackerpilot) [inline src/containers/slist.d]: `front` has a contract but this doesn't.

File: src/containers/treemap.d [hunk 3]
```diff
@@ -41,28 +46,81 @@ struct TreeMap(K, V, Allocator = Mallocator, alias less = "a < b",
...
 	{
 		alias CET = ContainerElementType!(This, V);
 		auto tme = TreeMapElement(key);
 		return cast(CET) tree.equalRange(tme).front.value;
 	}
 
+	/**
+	 * Returns: the value associated with the given key, or the given `defaultValue`.
+	 */
+	auto get(this This)(K key, lazy V defaultValue) inout @trusted
```
Comment (Hackerpilot) [inline src/containers/treemap.d]: `key` should probably be `const` here.

File: src/containers/treemap.d [hunk 4]
```diff
@@ -41,28 +46,81 @@ struct TreeMap(K, V, Allocator = Mallocator, alias less = "a < b",
...
+
+	/**
+	 * If the given key does not exist in the TreeMap, adds it with
+	 * the value `defaultValue`.
+	 *
+	 * Params:
+	 *     key = the key to look up
+	 *     value = the default value
+	 */
+	auto getOrAdd(this This)(K key, lazy V value) @safe
```
Comment (Hackerpilot) [inline src/containers/treemap.d]: Same here.

File: src/containers/treemap.d [hunk 5]
```diff
@@ -72,48 +130,92 @@ struct TreeMap(K, V, Allocator = Mallocator, alias less = "a < b",
 		return tree.remove(tme);
 	}
 
-	/// Returns: true if the mapping contains the given key
-	bool containsKey(K key)
+	/**
+	 * Returns: true if the mapping contains the given key
+	 */
+	bool containsKey(K key) inout pure nothrow @nogc @safe
```
Comment (Hackerpilot) [inline src/containers/treemap.d]: And here.

File: src/containers/unrolledlist.d [hunk 6]
```diff
@@ -273,7 +294,7 @@ struct UnrolledList(T, Allocator = Mallocator,
 		import containers.internal.backwards : bsf;
 		import std.string: format;
 		size_t index = bsf(_front.registry);
-		assert (index < nodeCapacity, format("%d", index));
+		//assert (index < nodeCapacity, format("%d", index));
```
Comment (Hackerpilot) [inline src/containers/unrolledlist.d]: Delete or move to an `in` contract.

File: src/containers/treemap.d [hunk 7]
```diff
@@ -41,28 +46,81 @@ struct TreeMap(K, V, Allocator = Mallocator, alias less = "a < b",
...
+	}
 
-	/// Supports $(B treeMap[key]) syntax.
-	auto opIndex(this This)(K key) const
+	/**
+	 * Supports $(B treeMap[key]) syntax.
+	 *
+	 * Throws: RangeError if the key does not exist.
+	 */
+	auto opIndex(this This)(K key) inout
```
Comment (Hackerpilot) [inline src/containers/treemap.d]: `const K`

File: src/containers/ttree.d [hunk 8]
```diff
@@ -54,73 +60,103 @@ struct TTree(T, Allocator = Mallocator, bool allowDuplicates = false,
...
+	 *         are both false) to `value`. This is useful in cases where opCmp
+	 *         and opEquals for `T` type have different meanings. For example,
+	 *         if the element type is a circle that has a position and a color,
+	 *         the circle could implement `opCmp` to sort by color, and calling
+	 *         `insert` with `overwrite` set to `true` would allow you to update
+	 *         the position of the circle with a certain color in the tree.
+	 * Returns: the number of values added.
 	 */
-	bool insert(T value)
+	size_t insert(T value, bool overwrite = false) @safe
```
Comment (Hackerpilot) [inline src/containers/ttree.d]: Should other containers also have overloads of `insert` that take ranges and return size_t?

File: src/containers/ttree.d [hunk 9]
```diff
@@ -133,74 +169,113 @@ struct TTree(T, Allocator = Mallocator, bool allowDuplicates = false,
 	 */
 	bool remove(T value, void delegate(T) cleanup = null)
```
Comment (Hackerpilot) [inline src/containers/ttree.d]: Figure out what should be done here when duplicates are enabled. This should probably return `size_t` like `insert`.

File: src/containers/ttree.d [hunk 10]
```diff
@@ -387,25 +472,42 @@ private:
...
+
 	// If we're storing a struct that defines opCmp, don't compare pointers as
 	// that is almost certainly not what the user intended.
-	static if (is(typeof(less) == string ) && less == "a < b"
-			&& isPointer!T && __traits(hasMember, PointerTarget!T, "opCmp"))
-		alias _less = binaryFun!"a.opCmp(*b) < 0";
+	static if (is(typeof(less) == string ))
+	{
+
+		static if (less == "a < b" && isPointer!T
```
Comment (Hackerpilot) [inline src/containers/ttree.d]: Document why this terrible, horrible hack is needed.

File: src/containers/ttree.d (hunk 8 above)
Comment (Hackerpilot) [reply src/containers/ttree.d]: Putting this off until later: https://github.com/dlang-community/containers/issues/105

File: src/containers/ttree.d (hunk 9 above)
Comment (Hackerpilot) [reply src/containers/ttree.d]: * Arbitrarily decided that `remove` would only remove a single element at a time
* documented this behavior 
* added a test

**Pull request merged**